# ML Zoomcamp Homework Week 01

## Import Libraries

In [2]:
import pandas as pd
import numpy as np

## Q1

In [3]:
pd.__version__

'2.2.2'

### Q1 Answer

`2.2.2`

## Getting the data

In [4]:
%%bash
wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv

--2024-09-20 19:15:31--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.108.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 298573 (292K) [text/plain]
Saving to: ‘laptops.csv.1’

     0K .......... .......... .......... .......... .......... 17% 1.27M 0s
    50K .......... .......... .......... .......... .......... 34% 2.08M 0s
   100K .......... .......... .......... .......... .......... 51% 2.58M 0s
   150K .......... .......... .......... .......... .......... 68% 3.48M 0s
   200K .......... .......... .......... .......... .......... 85% 3.38M 0s
   250K .......... .......... .......... .......... .         100% 5.70M=0.1s

2024-09-20 19:15:31 (2.45 MB/s) - ‘laptops.csv.1’ saved [298573/298573]



### Read in dataframe

In [5]:
df = pd.read_csv('laptops.csv')
df.head()

Laptop Status   Brand  \
0  ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...    New    Asus   
1  Alurin Go Start Intel Celeron N4020/8GB/256GB ...    New  Alurin   
2  ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...    New    Asus   
3  MSI Katana GF66 12UC-082XES Intel Core i7-1270...    New     MSI   
4  HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...    New      HP   

        Model            CPU  RAM  Storage Storage type       GPU  Screen  \
0  ExpertBook  Intel Core i5    8      512          SSD       NaN    15.6   
1          Go  Intel Celeron    8      256          SSD       NaN    15.6   
2  ExpertBook  Intel Core i3    8      256          SSD       NaN    15.6   
3      Katana  Intel Core i7   16     1000          SSD  RTX 3050    15.6   
4         15S  Intel Core i5   16      512          SSD       NaN    15.6   

  Touch  Final Price  
0    No      1009.00  
1    No       299.00  
2    No       789.00  
3    No      1199.00  
4    No       669.01

## Q2

In [11]:
print(df.shape[0])

2160


### Q2 Answer

There are 2160 records in this dataset

## Q3

In [15]:
df['Brand'].nunique()

27

### Q3 Answer

There are 27 laptop brands in the dataset

## Q4

In [141]:
# df.info()
len([col for col in df.columns if df[col].isnull().any()])

3

### Q4 Answers
There are 3 columns which are missing values

## Q5

In [142]:
# df[df['Brand'] == "Dell"].Model.unique()
print(df[df['Brand'] == "Dell"]["Final Price"].max())


3936.0


### Q5 Answer

The maximum final price of Dell notebooks in the dataset is `3936`

## Q6

In [143]:
screen_median = df['Screen'].median()
most_frequent_screen_val = df['Screen'].mode()
df['Screen'].fillna(most_frequent_screen_val)
print(screen_median)
new_screen_median = df['Screen'].median()
print(new_screen_median)

15.6
15.6


### Q6 Answer

The median has not changed after we used it to fill the missing values

## Q7

In [147]:
injoo_df = df[df['Brand'] == 'Innjoo'][['RAM', 'Storage', 'Screen']].copy()

In [151]:
X = injoo_df.to_numpy()

array([[  8. , 256. ,  15.6],
       [  8. , 512. ,  15.6],
       [  4. ,  64. ,  14.1],
       [  6. ,  64. ,  14.1],
       [  6. , 128. ,  14.1],
       [  6. , 128. ,  14.1]])

In [160]:
XTX = X.T.dot(X)

In [164]:
inv_XTX = np.linalg.inv(XTX)

In [165]:
y = [1100, 1300, 800, 900, 1000, 1100]

In [174]:
w = inv_XTX.dot(X.T).dot(y)

In [175]:
sum(w)

np.float64(91.2998806299555)